hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieMemoryConfig.java

```diff
@@ -140,7 +144,9 @@ private HoodieMemoryConfig() {
 
   public static String getDefaultSpillableMapBasePath() {
     String[] localDirs = FileIOUtils.getConfiguredLocalDirs();
-    return (localDirs != null && localDirs.length > 0) ? localDirs[0] : "/tmp/";
+    List<String> localDirLists = Arrays.asList(localDirs);
+    Collections.shuffle(localDirLists);
+    return !localDirLists.isEmpty() ? localDirLists.get(0) : "/tmp/";
```

Comment:
Nice catch ~

Reply:
> Nice catch ~

Nice catch +1, An optimization point, is it possible to return the path with the most available hard disk space?

Reply:
Yeah, I'm fine with the change if we can sift out the dist that has no enough disk space.

Reply:
Thank you very much for helping with the code review! `LOCAL_DIRS` is an environment variable set by YARN during Container startup, taking into consideration disk space availability and cases where a disk may be inaccessible. 

- Hudi#FileIOUtils#getYarnLocalDirs

```
  private static String getYarnLocalDirs() {
    String localDirs = System.getenv("LOCAL_DIRS");

    if (localDirs == null) {
      throw new HoodieIOException("Yarn Local dirs can't be empty");
    }
    return localDirs;
  }
```

- YARN

LinuxContainerExecutor#prepareContainer
```
builder.setExecutionAttribute(LOCALIZED_RESOURCES,
            ctx.getLocalizedResources())
        .setExecutionAttribute(USER, ctx.getUser())
        .setExecutionAttribute(CONTAINER_LOCAL_DIRS,
            ctx.getContainerLocalDirs())
        .setExecutionAttribute(CONTAINER_RUN_CMDS, ctx.getCommands())
        .setExecutionAttribute(CONTAINER_ID_STR,
            ctx.getContainer().getContainerId().toString());
```

ContainerLaunch#call
```

List<String> containerLocalDirs = getContainerLocalDirs(localDirs);

if (truststore != null) {
   addTruststoreVars(environment, containerWorkDir);
}

prepareContainer(localResources, containerLocalDirs);

// Write out the environment
exec.writeLaunchEnv(containerScriptOutStream, environment,
        localResources, launchContext.getCommands(), containerLogDir, user, nmEnvVars);
```

ContainerLaunch#prepareContainer
```
  protected List<String> getContainerLocalDirs(List<String> localDirs) {
    List<String> containerLocalDirs = new ArrayList<>(localDirs.size());
    String user = container.getUser();
    String appIdStr = app.getAppId().toString();
    String relativeContainerLocalDir = ContainerLocalizer.USERCACHE
        + Path.SEPARATOR + user + Path.SEPARATOR + ContainerLocalizer.APPCACHE
        + Path.SEPARATOR + appIdStr + Path.SEPARATOR;

    for (String localDir : localDirs) {
      containerLocalDirs.add(localDir + Path.SEPARATOR
          + relativeContainerLocalDir);
    }

    return containerLocalDirs;
  }
```

Reply:
- yarn.nodemanager.disk-health-checker.min-healthy-disks
```
The minimum fraction of number of disks to be healthy for the nodemanager to launch new containers. This applies to nm-local-dirs and nm-log-dirs.

 public static final String NM_MIN_HEALTHY_DISKS_FRACTION =
      NM_DISK_HEALTH_CHECK_PREFIX + "min-healthy-disks";
 public static final float DEFAULT_NM_MIN_HEALTHY_DISKS_FRACTION = 0.25F;
```
- yarn.nodemanager.disk-health-checker.max-disk-utilization-per-disk-percentage
```
The maximum percentage of disk space that can be used after which a disk is
marked as offline. Values can range from 0.0 to 100.0. If the value is
greater than or equal to 100, NM will check for full disk. This applies to
nm-local-dirs and nm-log-dirs.

public static final String NM_MAX_PER_DISK_UTILIZATION_PERCENTAGE =
      NM_DISK_HEALTH_CHECK_PREFIX + "max-disk-utilization-per-disk-percentage";
public static final float DEFAULT_NM_MAX_PER_DISK_UTILIZATION_PERCENTAGE = 90.0F;
```